readme.md
Kibotu-node
=============
![Build Status](https://github.com/kibotu/kibotu-node/actions/workflows/tests.yml/badge.svg)

This library provides many of the features in the official JavaScript kibotu library.  It is easy to use, and fully async. It is intended to be used on the server (it is not a client module). The in-browser client library is available
at [https://github.com/kibotu/kibotu-js](https://github.com/kibotu/kibotu-js).

Installation
------------

    npm install kibotu

Quick Start
-----------

```javascript
// grab the Kibotu factory
var Kibotu = require('kibotu');

// create an instance of the kibotu client
var kibotu = Kibotu.init('<YOUR_TOKEN>');

// initialize kibotu client configured to communicate over http instead of https
var kibotu = Kibotu.init('<YOUR_TOKEN>', {
    protocol: 'http',
});

// turn off keepAlive (reestablish connection on each request)
var kibotu = Kibotu.init('<YOUR_TOKEN>', {
    keepAlive: false,
});

// track an event with optional properties
kibotu.track('my event', {
    distinct_id: 'some unique client id',
    as: 'many',
    properties: 'as',
    you: 'want'
});
kibotu.track('played_game');

// set an IP address to get automatic geolocation info
kibotu.track('my event', {ip: '127.0.0.1'});

// track an event with a specific timestamp (up to 5 days old;
// use kibotu.import() for older events)
kibotu.track('timed event', {time: new Date()});

// create or update a user in Kibotu Engage
kibotu.people.set('billybob', {
    $first_name: 'Billy',
    $last_name: 'Bob',
    $created: (new Date('jan 1 2013')).toISOString(),
    plan: 'premium',
    games_played: 1,
    points: 0
});

// create or update a user in Kibotu Engage without altering $last_seen
// - pass option $ignore_time: true to prevent the $last_seen property from being updated
kibotu.people.set('billybob', {
    plan: 'premium',
    games_played: 1
}, {
    $ignore_time: true
});

// set a user profile's IP address to get automatic geolocation info
kibotu.people.set('billybob', {
    plan: 'premium',
    games_played: 1
}, {
    $ip: '127.0.0.1'
});

// set a user profile's latitude and longitude to get automatic geolocation info
kibotu.people.set('billybob', {
    plan: 'premium',
    games_played: 1
}, {
    $latitude: 40.7127753,
    $longitude: -74.0059728
});

// set a single property on a user
kibotu.people.set('billybob', 'plan', 'free');

// set a single property on a user, don't override
kibotu.people.set_once('billybob', 'first_game_play', (new Date('jan 1 2013')).toISOString());

// increment a numeric property
kibotu.people.increment('billybob', 'games_played');

// increment a numeric property by a different amount
kibotu.people.increment('billybob', 'points', 15);

// increment multiple properties
kibotu.people.increment('billybob', {'points': 10, 'games_played': 1});

// append value to a list
kibotu.people.append('billybob', 'awards', 'Great Player');

// append multiple values to a list
kibotu.people.append('billybob', {'awards': 'Great Player', 'levels_finished': 'Level 4'});

// merge value to a list (ignoring duplicates)
kibotu.people.union('billybob', {'browsers': 'ie'});

// merge multiple values to a list (ignoring duplicates)
kibotu.people.union('billybob', {'browsers': ['ie', 'chrome']});


// record a transaction for revenue analytics
kibotu.people.track_charge('billybob', 39.99);

// clear a users transaction history
kibotu.people.clear_charges('billybob');

// delete a user
kibotu.people.delete_user('billybob');

// delete a user in Kibotu Engage without altering $last_seen or resolving aliases
// - pass option $ignore_time: true to prevent the $last_seen property from being updated
// (useful if you subsequently re-import data for the same distinct ID)
kibotu.people.delete_user('billybob', {$ignore_time: true, $ignore_alias: true});

// Create an alias for an existing distinct id
kibotu.alias('distinct_id', 'your_alias');

// all functions that send data to kibotu take an optional
// callback as the last argument
kibotu.track('test', function(err) { if (err) throw err; });

// track multiple events at once
kibotu.track_batch([
    {
        event: 'recent event',
        properties: {
            time: new Date(),
            distinct_id: 'billybob',
            gender: 'male'
        }
    },
    {
        event: 'another recent event',
        properties: {
            distinct_id: 'billybob',
            color: 'red'
        }
    }
]);

// import an old event
var kibotu_importer = Kibotu.init('valid kibotu token', {
    secret: 'valid api secret for project'
});

// needs to be in the system once for it to show up in the interface
kibotu_importer.track('old event', { gender: '' });

kibotu_importer.import('old event', new Date(2012, 4, 20, 12, 34, 56), {
    distinct_id: 'billybob',
    gender: 'male'
});

// import multiple events at once
kibotu_importer.import_batch([
    {
        event: 'old event',
        properties: {
            time: new Date(2012, 4, 20, 12, 34, 56),
            distinct_id: 'billybob',
            gender: 'male'
        }
    },
    {
        event: 'another old event',
        properties: {
            time: new Date(2012, 4, 21, 11, 33, 55),
            distinct_id: 'billybob',
            color: 'red'
        }
    }
]);
```

FAQ
---
**Where is `kibotu.identify()`?**

`kibotu-node` is a server-side library, optimized for stateless shared usage; e.g.,
in a web application, the same kibotu instance is used across requests for all users.
Rather than setting a `distinct_id` through `identify()` calls like Kibotu client-side
libraries (where a single Kibotu instance is tied to a single user), this library
requires you to pass the `distinct_id` with every tracking call. See
https://github.com/kibotu/kibotu-node/issues/13.

**How do I get or set superproperties?**

See the previous answer: the library does not maintain user state internally and so has
no concept of superproperties for individual users. If you wish to preserve properties
for users between requests, you will need to load these properties from a source specific
to your app (e.g., your session store or database) and pass them explicitly with each
tracking call.


Tests
-----

    # in the kibotu directory
    npm install
    npm test

Alternative Clients and Related Tools
-------------------------------------

- [Kibotu-CLI](https://github.com/FGRibreau/kibotu-cli) - CLI for Kibotu API (currently only supports tracking functions)
- [Kibotu Data Export](https://github.com/michaelcarter/kibotu-data-export-js) - Supports various query and data-management APIs; runs in both Node.js and browser
- [Kibotu Data Export (strawbrary)](https://github.com/strawbrary/kibotu-data-export-js) - Fork of previous library, optimized for Node.js with support for streaming large raw exports

Attribution/Credits
-------------------

Heavily inspired by the original js library copyright Kibotu, Inc.
(http://kibotu.ai/)

Copyright (c) 2014-21 Kibotu
Original Library Copyright (c) 2012-14 Carl Sverre

Contributions from:
 - [Andres Gottlieb](https://github.com/andresgottlieb)
 - [Ken Perkins](https://github.com/kenperkins)
 - [Nathan Rajlich](https://github.com/TooTallNate)
 - [Thomas Watson Steen](https://github.com/watson)
 - [Gabor Ratky](https://github.com/rgabo)
 - [wwlinx](https://github.com/wwlinx)
 - [PierrickP](https://github.com/PierrickP)
 - [lukapril](https://github.com/lukapril)
 - [sandinmyjoints](https://github.com/sandinmyjoints)
 - [Jyrki Laurila](https://github.com/jylauril)
 - [Zeevl](https://github.com/zeevl)
 - [Tobias Baunbæk](https://github.com/freeall)
 - [Eduardo Sorribas](https://github.com/sorribas)
 - [Nick Chang](https://github.com/maeldur)
 - [Michael G](https://github.com/gmichael225)
 - [Tejas Manohar](https://github.com/tejasmanohar)
 - [Eelke Boezeman](https://github.com/godspeedelbow)
 - [Jim Thomas](https://github.com/Left47)
 - [Frank Chiang](https://github.com/chiangf)
 - [Morgan Croney](https://github.com/cruzanmo)
 - [Cole Furfaro-Strode](https://github.com/colestrode)
 - [Jonas Hermsmeier](https://github.com/jhermsmeier)
 - [Marko Klopets](https://github.com/mklopets)
 - [Cameron Diver](https://github.com/CameronDiver)
 - [veerabio](https://github.com/veerabio)
 - [Will Neild](https://github.com/wneild)
 - [Elijah Insua](https://github.com/tmpvar)
 - [Arsal Imam](https://github.com/ArsalImam)

License
-------------------

Released under the MIT license.  See file called LICENSE for more
details.


history.md
0.17.0 / 2022-08-11
==================
* support sending timestamps with millisecond precision

0.16.0 / 2022-06-02
==================
* support automatic geolocation with `geolocate` option (thanks tmpvar)
* send library version as property with events (thanks ArsalImam)

0.15.0 / 2022-05-20
==================
* use keepAlive by default for requests

0.14.0 / 2021-10-29
==================
* support $latitude and $longitude in profile operations (thanks wneild)

0.13.0 / 2020-09-04
==================
* support API Secret auth for imports and deprecate use of API Key

0.12.0 / 2020-08-31
==================
* https-proxy-agent upgrade to 5.0.0 to fix https.request patching and many subdependency upgrades (thanks veerabio)
* dropped support for node 8

0.11.0 / 2019-11-26
==================
* add support for Groups API

0.10.3 / 2019-10-09
==================
* upgrade https-proxy-agent for security fix (thanks omrilotan)

0.10.2 / 2019-03-26
==================
* type definitions for people.unset (thanks bradleyayers)

0.10.1 / 2018-12-03
==================
* support configurable API path (thanks CameronDiver)

0.9.2 / 2018-05-22
==================
* add type declarations file (thanks mklopets)

0.9.1 / 2018-04-12
==================
* upgrade https-proxy-agent for security fix

0.9.0 / 2018-02-09
==================
* default to tracking over HTTPS (thanks jhermsmeier)

0.8.0 / 2017-11-28
==================
* upgraded node-https-proxy-agent to v2.1.1 for security patch (see
https://github.com/TooTallNate/node-https-proxy-agent/issues/37)

0.7.0 / 2017-04-07
===================
* added `track_batch` for tracking multiple recent events per request (thanks cruzanmo)
* support for routing requests through proxy server specified in env var `HTTPS_PROXY`
or `HTTP_PROXY` (thanks colestrode)
* dropped support for node 0.10 and 0.12

0.6.0 / 2017-01-03
===================
* support for `time` field in `kibotu.track()` (thanks cruzanmo)

0.5.0 / 2016-09-15
===================
* optional https support (thanks chiangf)

0.4.1 / 2016-09-09
===================
* include `$ignore_alias` in permitted `people` modifiers (thanks Left47)

0.4.0 / 2016-02-09
===================
* allow optional `modifiers` in all `people` calls for `$ignore_time`, `$ip`,
  and `$time` fields

0.3.2 / 2015-12-10
===================
* correct `$delete` field in `people.delete_user` request (thanks godspeedelbow)

0.3.1 / 2015-08-06
===================
* added config option for API host (thanks gmichael225)

0.3.0 / 2015-08-06
===================
* added people.union support (thanks maeldur)

0.2.0 / 2015-04-14
===================
* added batch import support

0.1.1 / 2015-03-27
===================
* fixed callback behavior in track_charges when no properties supplied
(thanks sorribas)

0.1.0 / 2015-03-20
===================
* updated URL metadata (thanks freeall)
* updated dev dependencies
* added builds for iojs, node 0.12, dropped support for node <0.10

0.0.20 / 2014-05-11
====================
* removed hardcoded port 80 for more flexibility (thanks zeevl)

0.0.19 / 2014.04.03
====================
* added people.append (thanks jylauril)

0.0.18 / 2013-08-23
====================
* added callback to alias (thanks to sandinmyjoints)
* added verbose config option (thanks to sandinmyjoints)
* added unset method (thanks to lukapril)

0.0.17 / 2013-08-12
====================
* added alias method (thanks to PierrickP)

0.0.16 / 2013-06-29
====================
* allow special key "ip" to be 0 in people.set (thanks to wwlinx)

0.0.15 / 2013-05-24
====================
* adds set once functionality to people (thanks to avoid3d)
* $ignore_time in people.set (thanks to Rick Cotter)

0.0.14 / 2013-03-28
====================
* revert Randal's http only patch since Kibotu indeed supports https.
* handles the ip property in a property object properly for people calls

0.0.13 / 2013-03-25
====================
* force requests to go over http [reverted in 0.0.14]

0.0.12 / 2013-01-24
====================
* track_charge() no longer includes $time by default, rather it lets
Kibotu's servers set the time when they receive the transaction.  This
doesn't modify the ability for the user to pass in their own $time (for
importing transactions).

0.0.11 / 2013-01-11
====================
* added track_charge() method which provides the ability to record user
transactions for revenue analytics.
* added clear_charges() method which provides the ability to remove a
users transactions from Kibotu
* added tests for delete_user()

0.0.10 / 2012-11-26
====================
* added import() method which provides the ability to import events
older than 5 days.  Contributions from Thomas Watson Steen.

0.0.9 / 2012-11-15
===================
* removed time from properties sent to server.  This is to ensure that
UTC is always used.  Kibotu will set the correct time as soon as they
receive the event.

0.0.8 / 2012-10-24
===================
* added mp_lib property, so people can segment by library

0.0.7 / 2012-01-05
===================
* added unit tests
* people.increment() only prints error message if debug is true

0.0.6 / 2012-01-01
===================
* added engage support
  * people.set()
  * people.increment()
  * people.delete_user()
* deprecated old constructor: require("kibotu").Client(token)
* added new constructor: require("kibotu").init(token)


test/alias.js
var Kibotu    = require('../lib/kibotu-node'),
    Sinon       = require('sinon');

exports.alias = {
    setUp: function(next) {
        this.kibotu = Kibotu.init('token', { key: 'key' });

        Sinon.stub(this.kibotu, 'send_request');

        next();
    },

    tearDown: function(next) {
        this.kibotu.send_request.restore();

        next();
    },

    "calls send_request with correct endpoint and data": function(test) {
        var alias = "test",
            distinct_id = "old_id",
            expected_endpoint = "/track",
            expected_data = {
                event: '$create_alias',
                properties: {
                    distinct_id: distinct_id,
                    alias: alias,
                    token: 'token'
                }
            };

        this.kibotu.alias(distinct_id, alias);

        test.ok(
            this.kibotu.send_request.calledWithMatch({ endpoint: expected_endpoint, data: expected_data }),
            "alias didn't call send_request with correct arguments"
        );

        test.done();
    }
};


test/track.js
var proxyquire  = require('proxyquire'),
    Sinon       = require('sinon'),
    https       = require('https'),
    events      = require('events'),
    Kibotu    = require('../lib/kibotu-node'),
    packageInfo = require('../package.json');

var mock_now_time = new Date(2016, 1, 1).getTime();

exports.track = {
    setUp: function(next) {
        this.kibotu = Kibotu.init('token');
        this.clock = Sinon.useFakeTimers(mock_now_time);

        Sinon.stub(this.kibotu, 'send_request');

        next();
    },

    tearDown: function(next) {
        this.kibotu.send_request.restore();
        this.clock.restore();

        next();
    },

    "calls send_request with correct endpoint and data": function(test) {
        var event = "test",
            props = { key1: 'val1' },
            expected_endpoint = "/track",
            expected_data = {
                event: 'test',
                properties: {
                    key1: 'val1',
                    token: 'token'
                }
            };

        this.kibotu.track(event, props);

        test.ok(
            this.kibotu.send_request.calledWithMatch({ endpoint: expected_endpoint, data: expected_data }),
            "track didn't call send_request with correct arguments"
        );

        test.done();
    },

    "can be called with optional properties": function(test) {
        var expected_endpoint = "/track",
            expected_data = {
                event: 'test',
                properties: {
                    token: 'token'
                }
            };

        this.kibotu.track("test");

        test.ok(
            this.kibotu.send_request.calledWithMatch({ endpoint: expected_endpoint, data: expected_data }),
            "track didn't call send_request with correct arguments"
        );
        test.done();
    },

    "can be called with optional callback": function(test) {
        var expected_endpoint = "/track",
            expected_data = {
                event: 'test',
                properties: {
                    token: 'token'
                }
            };

        this.kibotu.send_request.callsArgWith(1, undefined);

        test.expect(1);
        this.kibotu.track("test", function(e) {
            test.equal(e, undefined, "error should be undefined");
            test.done();
        });
    },

    "supports Date object for time": function(test) {
        var event = 'test',
            time = new Date(mock_now_time),
            props = { time: time },
            expected_endpoint = "/track",
            expected_data = {
                event: 'test',
                properties: {
                    token: 'token',
                    time: time.getTime(),
                    mp_lib: 'node',
                    $lib_version: packageInfo.version
                }
            };

        this.kibotu.track(event, props);

        test.ok(
            this.kibotu.send_request.calledWithMatch({ endpoint: expected_endpoint, data: expected_data }),
            "track didn't call send_request with correct arguments"
        );
        test.done();
    },

    "supports unix timestamp for time": function(test) {
        var event = 'test',
            time = mock_now_time,
            props = { time: time },
            expected_endpoint = "/track",
            expected_data = {
                event: 'test',
                properties: {
                    token: 'token',
                    time: time,
                    mp_lib: 'node',
                    $lib_version: packageInfo.version
                }
            };

        this.kibotu.track(event, props);

        test.ok(
            this.kibotu.send_request.calledWithMatch({ endpoint: expected_endpoint, data: expected_data }),
            "track didn't call send_request with correct arguments"
        );
        test.done();
    },

    "throws error if time is not a number or Date": function(test) {
        var event = 'test',
            props = { time: 'not a number or Date' };

        test.throws(
            this.kibotu.track.bind(this, event, props),
            /`time` property must be a Date or Unix timestamp/,
            "track didn't throw an error when time wasn't a number or Date"
        );
        test.done();
    },

    "does not require time property": function(test) {
        var event = 'test',
            props = {};

        test.doesNotThrow(this.kibotu.track.bind(this, event, props));
        test.done();
    }
};

exports.track_batch = {
    setUp: function(next) {
        this.kibotu = Kibotu.init('token');
        this.clock = Sinon.useFakeTimers();
        Sinon.stub(this.kibotu, 'send_request');
        next();
    },

    tearDown: function(next) {
        this.kibotu.send_request.restore();
        this.clock.restore();
        next();
    },

    "calls send_request with correct endpoint, data, and method": function(test) {
        var expected_endpoint = "/track",
            event_list = [
                {event: 'test',  properties: {key1: 'val1', time: 500 }},
                {event: 'test',  properties: {key2: 'val2', time: 1000}},
                {event: 'test2', properties: {key2: 'val2', time: 1500}}
            ],
            expected_data = [
                {event: 'test',  properties: {key1: 'val1', time: 500,  token: 'token'}},
                {event: 'test',  properties: {key2: 'val2', time: 1000, token: 'token'}},
                {event: 'test2', properties: {key2: 'val2', time: 1500, token: 'token'}}
            ];

        this.kibotu.track_batch(event_list);

        test.ok(
            this.kibotu.send_request.calledWithMatch({
                method: "POST",
                endpoint: expected_endpoint,
                data: expected_data
            }),
            "track_batch didn't call send_request with correct arguments"
        );

        test.done();
    },

    "does not require the time argument for every event": function(test) {
        var event_list = [
            {event: 'test',  properties: {key1: 'val1', time: 500 }},
            {event: 'test',  properties: {key2: 'val2', time: 1000}},
            {event: 'test2', properties: {key2: 'val2'            }}
        ];
        test.doesNotThrow(this.kibotu.track_batch.bind(this, event_list));
        test.done();
    },

    "batches 50 events at a time": function(test) {
        var event_list = [];
        for (var ei = 0; ei < 130; ei++) { // 3 batches: 50 + 50 + 30
            event_list.push({event: 'test',  properties: {key1: 'val1', time: 500 + ei }});
        }

        this.kibotu.track_batch(event_list);

        test.equals(
            3, this.kibotu.send_request.callCount,
            "track_batch didn't call send_request correct number of times"
        );

        test.done();
    }
};

exports.track_batch_integration = {
    setUp: function(next) {
        this.kibotu = Kibotu.init('token', { key: 'key' });
        this.clock = Sinon.useFakeTimers();

        Sinon.stub(https, 'request');

        this.http_emitter = new events.EventEmitter();

        // stub sequence of https responses
        this.res = [];
        for (var ri = 0; ri < 5; ri++) {
            this.res.push(new events.EventEmitter());
            https.request.onCall(ri).callsArgWith(1, this.res[ri]);
            https.request.onCall(ri).returns({
                write: function () {},
                end: function () {},
                on: function(event) {}
            });
        }

        this.event_list = [];
        for (var ei = 0; ei < 130; ei++) { // 3 batches: 50 + 50 + 30
            this.event_list.push({event: 'test',  properties: {key1: 'val1', time: 500 + ei }});
        }

        next();
    },

    tearDown: function(next) {
        https.request.restore();
        this.clock.restore();

        next();
    },

    "calls provided callback after all requests finish": function(test) {
        test.expect(2);
        this.kibotu.track_batch(this.event_list, function(error_list) {
            test.equals(
                3, https.request.callCount,
                "track_batch didn't call send_request correct number of times before callback"
            );
            test.equals(
                null, error_list,
                "track_batch returned errors in callback unexpectedly"
            );
            test.done();
        });
        for (var ri = 0; ri < 3; ri++) {
            this.res[ri].emit('data', '1');
            this.res[ri].emit('end');
        }
    },

    "passes error list to callback": function(test) {
        test.expect(1);
        this.kibotu.track_batch(this.event_list, function(error_list) {
            test.equals(
                3, error_list.length,
                "track_batch didn't return errors in callback"
            );
            test.done();
        });
        for (var ri = 0; ri < 3; ri++) {
            this.res[ri].emit('data', '0');
            this.res[ri].emit('end');
        }
    },

    "calls provided callback when options are passed": function(test) {
        test.expect(2);
        this.kibotu.track_batch(this.event_list, {max_batch_size: 100}, function(error_list) {
            test.equals(
                3, https.request.callCount,
                "track_batch didn't call send_request correct number of times before callback"
            );
            test.equals(
                null, error_list,
                "track_batch returned errors in callback unexpectedly"
            );
            test.done();
        });
        for (var ri = 0; ri < 3; ri++) {
            this.res[ri].emit('data', '1');
            this.res[ri].emit('end');
        }
    },

    "sends more requests when max_batch_size < 50": function(test) {
        test.expect(2);
        this.kibotu.track_batch(this.event_list, {max_batch_size: 30}, function(error_list) {
            test.equals(
                5, https.request.callCount, // 30 + 30 + 30 + 30 + 10
                "track_batch didn't call send_request correct number of times before callback"
            );
            test.equals(
                null, error_list,
                "track_batch returned errors in callback unexpectedly"
            );
            test.done();
        });
        for (var ri = 0; ri < 5; ri++) {
            this.res[ri].emit('data', '1');
            this.res[ri].emit('end');
        }
    },

    "can set max concurrent requests": function(test) {
        var async_all_stub = Sinon.stub();
        var PatchedKibotu = proxyquire('../lib/kibotu-node', {
            './utils': {async_all: async_all_stub},
        });
        async_all_stub.callsArgWith(2, null);
        this.kibotu = PatchedKibotu.init('token', { key: 'key' });

        test.expect(2);
        this.kibotu.track_batch(this.event_list, {max_batch_size: 30, max_concurrent_requests: 2}, function(error_list) {
            // should send 5 event batches over 3 request batches:
            // request batch 1: 30 events, 30 events
            // request batch 2: 30 events, 30 events
            // request batch 3: 10 events
            test.equals(
                3, async_all_stub.callCount,
                "track_batch didn't batch concurrent https requests correctly"
            );
            test.equals(
                null, error_list,
                "track_batch returned errors in callback unexpectedly"
            );
            test.done();
        });
        for (var ri = 0; ri < 3; ri++) {
            this.res[ri].emit('data', '1');
            this.res[ri].emit('end');
        }
    },

    "behaves well without a callback": function(test) {
        test.expect(2);
        this.kibotu.track_batch(this.event_list);
        test.equals(
            3, https.request.callCount,
            "track_batch didn't call send_request correct number of times"
        );
        this.kibotu.track_batch(this.event_list, {max_batch_size: 100});
        test.equals(
            5, https.request.callCount, // 3 + 100 / 50; last request starts async
            "track_batch didn't call send_request correct number of times"
        );
        test.done();
    }
};




test/config.js
var Kibotu = require('../lib/kibotu-node');

exports.config = {
    setUp: function(cb) {
        this.kibotu = Kibotu.init('asjdf');
        cb();
    },

    "is set to correct defaults": function(test) {
        test.deepEqual(this.kibotu.config, {
            test: false,
            debug: false,
            verbose: false,
            host: 'api.kibotu.ai',
            protocol: 'https',
            path: '',
            keepAlive: true,
            geolocate: false,
        }, "default config is incorrect");
        test.done();
    },

    "is modified by set_config": function(test) {
        test.equal(this.kibotu.config.test, false, "default config has incorrect value for test");

        this.kibotu.set_config({ test: true });

        test.equal(this.kibotu.config.test, true, "set_config failed to modify the config");

        test.done();
    },

    "can be set during init": function(test) {
        var mp = Kibotu.init('token', { test: true });

        test.equal(mp.config.test, true, "init() didn't set the config correctly");
        test.done();
    },

    "host config is split into host and port": function(test) {
        const exampleHost = 'api.example.com';
        const examplePort = 70;
        const hostWithoutPortConfig = Kibotu.init('token', {host: exampleHost}).config;
        test.equal(hostWithoutPortConfig.port, undefined, "port should not have been added to config");
        test.equal(hostWithoutPortConfig.host, exampleHost, `host should match ${exampleHost}`);

        const hostWithPortConfig = Kibotu.init('token', {host: `${exampleHost}:${examplePort}`}).config;
        test.equal(hostWithPortConfig.port, examplePort, "port should have been added to config");
        test.equal(hostWithPortConfig.host, exampleHost, `host should match ${exampleHost}`);

        test.done();
    },
};


lib/groups.js
/**
 * Group profile methods. Learn more: https://help.kibotu.ai/hc/en-us/articles/360025333632
 */

const {ProfileHelpers} = require('./profile_helpers');

class KibotuGroups extends ProfileHelpers() {
    constructor(mp_instance) {
        super();
        this.kibotu = mp_instance;
        this.endpoint = '/groups';
    }

    /** groups.set_once(group_key, group_id, prop, to, modifiers, callback)
        ---
        The same as groups.set, but adds a property value to a group only if it has not been set before.
    */
    set_once(group_key, group_id, prop, to, modifiers, callback) {
        const identifiers = {$group_key: group_key, $group_id: group_id};
        this._set(prop, to, modifiers, callback, {identifiers, set_once: true});
    }

    /**
        groups.set(group_key, group_id, prop, to, modifiers, callback)
        ---
        set properties on a group profile

        usage:

            kibotu.groups.set('company', 'Acme Inc.', '$name', 'Acme Inc.');

            kibotu.groups.set('company', 'Acme Inc.', {
                'Industry': 'widgets',
                '$name': 'Acme Inc.',
            });
    */
    set(group_key, group_id, prop, to, modifiers, callback) {
        const identifiers = {$group_key: group_key, $group_id: group_id};
        this._set(prop, to, modifiers, callback, {identifiers});
    }

    /**
     groups.delete_group(group_key, group_id, modifiers, callback)
        ---
        delete a group profile permanently

        usage:

        kibotu.groups.delete_group('company', 'Acme Inc.');
        */
    delete_group(group_key, group_id, modifiers, callback) {
        const identifiers = {$group_key: group_key, $group_id: group_id};
        this._delete_profile({identifiers, modifiers, callback});
    }

    /**
     groups.remove(group_key, group_id, data, modifiers, callback)
        ---
        remove a value from a list-valued group profile property.

        usage:

        kibotu.groups.remove('company', 'Acme Inc.', {'products': 'anvil'});

        kibotu.groups.remove('company', 'Acme Inc.', {
            'products': 'anvil',
            'customer segments': 'coyotes'
        });
        */
    remove(group_key, group_id, data, modifiers, callback) {
        const identifiers = {$group_key: group_key, $group_id: group_id};
        this._remove({identifiers, data, modifiers, callback});
    }

    /**
     groups.union(group_key, group_id, data, modifiers, callback)
        ---
        merge value(s) into a list-valued group profile property.

        usage:

        kibotu.groups.union('company', 'Acme Inc.', {'products': 'anvil'});

        kibotu.groups.union('company', 'Acme Inc.', {'products': ['anvil'], 'customer segments': ['coyotes']});
        */
    union(group_key, group_id, data, modifiers, callback) {
        const identifiers = {$group_key: group_key, $group_id: group_id};
        this._union({identifiers, data, modifiers, callback})
    }

    /**
     groups.unset(group_key, group_id, prop, modifiers, callback)
        ---
        delete a property on a group profile

        usage:

        kibotu.groups.unset('company', 'Acme Inc.', 'products');

        kibotu.groups.unset('company', 'Acme Inc.', ['products', 'customer segments']);
        */
    unset(group_key, group_id, prop, modifiers, callback) {
        const identifiers = {$group_key: group_key, $group_id: group_id};
        this._unset({identifiers, prop, modifiers, callback})
    }
}

exports.KibotuGroups = KibotuGroups;


example.js
// grab the Kibotu factory
var Kibotu = require('./lib/kibotu-node');

// create an instance of the kibotu client
var kibotu = Kibotu.init('962dbca1bbc54701d402c94d65b4a20e');
kibotu.set_config({ debug: true });

// track an event with optional properties
kibotu.track("my event", {
    distinct_id: "some unique client id",
    as: "many",
    properties: "as",
    you: "want"
});
kibotu.track("played_game");

// create or update a user in Kibotu Engage
kibotu.people.set("billybob", {
    $first_name: "Billy",
    $last_name: "Bob",
    $created: (new Date('jan 1 2013')).toISOString(),
    plan: "premium",
    games_played: 1,
    points: 0
});

// create or update a user in Kibotu Engage without altering $last_seen
// - pass option `$ignore_time: true` to prevent the $last_seen property from being updated
kibotu.people.set("billybob", {
    plan: "premium",
    games_played: 1
}, {
    $ignore_time: true
});

// set a single property on a user
kibotu.people.set("billybob", "plan", "free");

// set a single property on a user, don't override
kibotu.people.set_once("billybob", "first_game_play", (new Date('jan 1 2013')).toISOString());

// increment a numeric property
kibotu.people.increment("billybob", "games_played");

// increment a numeric property by a different amount
kibotu.people.increment("billybob", "points", 15);

// increment multiple properties
kibotu.people.increment("billybob", {"points": 10, "games_played": 1});

// append value to a list
kibotu.people.append("billybob", "awards", "Great Player");

// append multiple values to a list
kibotu.people.append("billybob", {"awards": "Great Player", "levels_finished": "Level 4"});

// record a transaction for revenue analytics
kibotu.people.track_charge("billybob", 39.99);

// clear a users transaction history
kibotu.people.clear_charges("billybob");

// delete a user
kibotu.people.delete_user("billybob");

// all functions that send data to kibotu take an optional
// callback as the last argument
kibotu.track("test", function(err) { if (err) { throw err; } });

// import an old event
var kibotu_importer = Kibotu.init('valid kibotu token', {
    secret: "valid api secret for project"
});
kibotu_importer.set_config({ debug: true });

// needs to be in the system once for it to show up in the interface
kibotu_importer.track('old event', { gender: '' });

kibotu_importer.import("old event", new Date(2012, 4, 20, 12, 34, 56), {
    distinct_id: 'billybob',
    gender: 'male'
});

// import multiple events at once
kibotu_importer.import_batch([
    {
        event: 'old event',
        properties: {
            time: new Date(2012, 4, 20, 12, 34, 56),
            distinct_id: 'billybob',
            gender: 'male'
        }
    },
    {
        event: 'another old event',
        properties: {
            time: new Date(2012, 4, 21, 11, 33, 55),
            distinct_id: 'billybob',
            color: 'red'
        }
    }
]);
